### PR TITLE
remove parser

### PR DESCRIPTION
**Pull request by Edoardo-Pedicillo**
https://github.com/qiboteam/boostvqe/issues/84#issuecomment-2415819149

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,12 +23,22 @@ The source code is located in `./src/boostvqe/.` and its composed of:
...
-```sh
-python main.py --help
+```py
+
+from boostvqe.boost import main
+from boostvqe.ansatze import hdw_efficient
+
+from qibo.models.dbi.double_bracket import DoubleBracketGeneratorType
+
+help(main) #If you need some info
```
Comment (andrea-pasquale) [inline README.md:37]: This is ok, however I don't like that when launching the script it will open a separate window that you need to close. Perhaps it is better just to drop it.

File: src/boostvqe/boost.py [hunk 2]
```diff
@@ -38,41 +37,135 @@
...
-def main(args):
-    """VQE training."""
-    # set backend and number of classical threads
 
-    if args.platform is not None:
-        qibo.set_backend(backend=args.backend, platform=args.platform)
-    else:
-        qibo.set_backend(backend=args.backend)
-        args.platform = GlobalBackend().platform
+def main(
```
Comment (andrea-pasquale) [inline src/boostvqe/boost.py:43]: At this point I will not call it `main` anymore.
Perhaps `boostvqe`, `boost` or even `dbqaxvqe`?

File: README.md [hunk 3]
```diff
@@ -8,11 +8,10 @@ Boosting variational eigenstate preparation algorithms limited by training and n
 The package can be installed by source after cloning the repository:
 
 ```sh
+cd boostvqe
```
Comment (andrea-pasquale) [inline README.md:11]: Here do we really need the `-e` option?

File: README.md [hunk 4]
```diff
@@ -24,12 +23,22 @@ The source code is located in `./src/boostvqe/.` and its composed of:
...
+from boostvqe.boost import main
+from boostvqe.ansatze import hdw_efficient
+
+from qibo.models.dbi.double_bracket import DoubleBracketGeneratorType
+
+help(main) #If you need some info
+
+circuit = hdw_efficient(nqubits=2, nlayers=2)
+
+main(circuit, "test_output", mode = DoubleBracketGeneratorType.group_commutator)
```
Comment (andrea-pasquale) [inline README.md:41]: Here I would put a comment explaining what is `test_output` and briefly also `mode`

File: src/boostvqe/boost.py [hunk 5]
```diff
@@ -38,41 +37,135 @@
...
+                optimizer,
+                nqubits,
+                seed,
+                decay_rate_lr,
+                nlayers,
+            )
+        )
+    )
+    ham = getattr(Model, hamiltonian)(nqubits)  # TODO : use only Model and not str
+    # ham = hamiltonian(nqubits)
```
Comment (andrea-pasquale) [inline src/boostvqe/boost.py:153]:
```suggestion
```

File: src/boostvqe/boost.py [hunk 6]
```diff
@@ -38,41 +37,135 @@
...
+            Number of shots (measurements) to use during the quantum computation.
+
+        nlayers (int):
+            Number of circuit layers.
+
+        mode:
+            Define the DBI Generator.
+    """
+
+    output_dict = locals()
```
Comment (andrea-pasquale) [inline src/boostvqe/boost.py:124]: I'm not a huge fan of `locals` but it does the job so we can keep it.

**Review by andrea-pasquale — APPROVED** [review]
Thanks @Edoardo-Pedicillo, just a few minor changes.

File: src/boostvqe/boost.py (hunk 6 above)
Comment (Edoardo-Pedicillo) [reply src/boostvqe/boost.py:124]: me too, but I was not able to find better alternatives

File: src/boostvqe/boost.py [hunk 7]
```diff
@@ -38,41 +37,130 @@
...
+            Custom options to fine-tune the behavior of the optimizer.
+
+        tol (float, default: TOL):
+            Absolute precision tolerance to stop VQE training once the desired convergence is achieved.
+
+        decay_rate_lr (float, default: 1.0):
+            Learning rate decay factor used when the optimizer is SGD (stochastic gradient descent).
+
+        nboost (int, default: 1):
+            Number of times DBI (Deterministic Boost Iteration) is applied in the optimization process.
```
Comment (marekgluza) [inline src/boostvqe/boost.py:89]: DBI =  double-bracket iteration

is this part of the CMA optimizer?

File: src/boostvqe/boost.py [hunk 8]
```diff
@@ -38,41 +37,130 @@
...
+            If set to 1, no boosting is performed.
+
+        boost_frequency (int, default: 10):
+            Number of optimization steps between DBI boosting calls.
 
-    if args.optimizer_options is None:
-        opt_options = {}
+        dbi_steps (int, default: 1):
```
Comment (marekgluza) [inline src/boostvqe/boost.py:95]:
```suggestion
        dbqa_steps (int, default: 1):
```
this code should be capable of running DBI (dense matrices) and GCI (qibo.Circuit)

File: src/boostvqe/boost.py [hunk 9]
```diff
@@ -38,41 +37,130 @@
...
+            Number of DBI iterations performed each time DBI is called.
+
+        stepsize (float, default: 0.01):
```
Comment (marekgluza) [inline src/boostvqe/boost.py:98]:
```suggestion
        dbr_duration (float, default: 0.01):
```
Each DBI step is DBR (double-bracket rotation) and in the strategies proceeding we always say DBR duration

File: src/boostvqe/boost.py [hunk 10]
```diff
@@ -38,41 +37,130 @@
...
+            Number of times DBI (Deterministic Boost Iteration) is applied in the optimization process.
+            If set to 1, no boosting is performed.
+
+        boost_frequency (int, default: 10):
+            Number of optimization steps between DBI boosting calls.
 
-    if args.optimizer_options is None:
-        opt_options = {}
+        dbi_steps (int, default: 1):
+            Number of DBI iterations performed each time DBI is called.
```
Comment (marekgluza) [inline src/boostvqe/boost.py:96]: iterations of what?

File: src/boostvqe/boost.py [hunk 11]
```diff
@@ -38,41 +37,130 @@
...
+
+        stepsize (float, default: 0.01):
+            Step size used during DBI updates.
+
+        store_h (bool, default: False):
+            If this flag is set, the Hamiltonian `H` is stored at each iteration.
+
+        hamiltonian (str, default: "XXZ"):
```
Comment (marekgluza) [inline src/boostvqe/boost.py:104]: after merging #77 this can become also a symbolic Hamiltonian

File: src/boostvqe/boost.py [hunk 12]
```diff
@@ -38,41 +37,130 @@
...
+
+        nshots (int, optional):
+            Number of shots (measurements) to use during the quantum computation.
+
+        nlayers (int):
+            Number of circuit layers.
+
+        mode:
+            Define the DBI Generator.
+    """
```
Comment (marekgluza) [inline src/boostvqe/boost.py:118]:
```suggestion
    """please_be_verbose:
    Flag which switches on verbose reporting - after each DBQA step a report is printed on energy gain, fidelity and circuit depth and stored in the output_folder as a log file.
```

File: README.md [hunk 13]
```diff
@@ -24,14 +23,24 @@ The source code is located in `./src/boostvqe/.` and its composed of:
...
+
+from boostvqe.boost import dbqa_vqe
+from boostvqe.ansatze import hdw_efficient
+
+from qibo.models.dbi.double_bracket import DoubleBracketGeneratorType
+
+circuit = hdw_efficient(nqubits=2, nlayers=2)
+output_folder = "output"
+dbqa_vqe(circuit, output_folder, mode = DoubleBracketGeneratorType.group_commutator)
 ```
```
Comment (marekgluza) [inline README.md:40]: here it would be nice to be able to print a report on how much was gained by running the dbqa_vqe

**Comment by marekgluza** [comment]
Thank you @Edoardo-Pedicillo !

I like the overall code structure and e.g. the clarity of the new Readme.

I removed Sam as reviewer:
- agree with @MatteoRobbiati and @andrea-pasquale if the coding approach works for you
- then merge to finish this refactoring
- I'll move ahead with the journal submission if you confirm that the code can reproduce what we had in plots 
- the README example is very small so we should have a simple documentation script/notebook which additionally explains how to extract the energy gain and circuit depth

**Comment by marekgluza** [comment]
Will this close https://github.com/qiboteam/boostvqe/issues/83 where the number of trotter steps and dbqa steps should be different?

File: README.md (hunk 13 above)
Comment (Edoardo-Pedicillo) [reply README.md:40]: I am not sure about this since this function should be called in other scripts. By the way, this PR is quite big, so this request should be handled in a separate PR.

Comment (marekgluza) [reply README.md:40]: yep, let's move ahead towards merge without it

File: src/boostvqe/boost.py (hunk 12 above)
Comment (Edoardo-Pedicillo) [reply src/boostvqe/boost.py:118]: Also this should be done in a separate PR

Comment (marekgluza) [reply src/boostvqe/boost.py:118]: Ok, let's move ahead without it
